First steps towards ROS2

rgov opened the pull request:
This branch contains some first steps towards migrating to ROS2.

ROS1 and Ubuntu 20.04 LTS reached end-of-life on May 31, 2025. While this does not present any immediate issue for us, we need to begin considering a roadmap towards supported releases.

I'm initially targeting ROS2 ~Humble~ Foxy, which supports Ubuntu 20.04, the same as ROS1 Noetic. This also aligns with our current Jetson release, r35. We can then upgrade smoothly to Ubuntu 22.04, Jetson r36, and ROS2 Humble or Iron.


## Migration Plan

A rough outline of this migration could look like this:

1. Move ROS1 packages into a new workspace, `ros1_ws`.
2. Introduce ROS2 packages in a new, separate workspace, `ros2_ws`.
3. Port all message definitions into the ROS2 workspace.
4. ~~Build the [`ros1_bridge`](https://github.com/ros2/ros1_bridge) using message definitions in both workspaces.~~
5. ~~Modify the ROS1 launch files to start the bridge: `ros2 run ros1_bridge dynamic_bridge`~~
6. Port packages one-by-one from ROS1 to ROS2.

In this sequence, ROS1 remains the "primary" ROS version, for example still using ROS1's `roslaunch` and `rosbag record`. We would eventually switch over to using ROS2 as the primary.


## Opportunities

- Split driver packages like AML, Digital Logger, JVL, and RBR off into separate repositories.
- Create a common node implementation pattern, possibly around [lifecycle nodes](https://design.ros2.org/articles/node_lifecycle.html) and `asyncio`.
- Migrate from launch XML files to Python scripts, possibly
    - obsoleting the `phyto-arm` script,
    - making configuration more intuitive, and
    - allowing nodes to respawn with less disruption to the whole system.

## Status

| Package Name                                          | Status      |
|-------------------------------------------------------|-------------|
| **INTERNAL PACKAGES** | |
| aml_ctd | ✳️ |
| dli_power_switch | ✳️ |
| ifcb | ✳️ |
| jvl_motor | ✳️ |
| phyto_arm |  |
| └ arm_base | not ported |
| └ arm_chanos | not ported |
| └ arm_ifcb | not ported |
| └ ifcb_runner | not ported |
| └ lock_manager | not ported |
| └ network_data_capture | ✳️ |
| └ profiler_node | not ported |
| └ seatrac_starlink_manager | not ported |
| └ web_node | ✳️ |
| └  winch_node | not ported |
| rbr_maestro3_ctd | partially ported |
| **EXTERNAL PACKAGES** | |
| [ros-launchpad](https://github.com/WHOIGit/ros-launchpad) | not ported, see https://github.com/WHOIGit/ros-launchpad/issues/1 |
| [ros-rtsp-camera](https://github.com/whoigit/ros-rtsp-camera) | ✅ replaced with ~~[rtsp-camera](https://github.com/mzahana/rtsp_camera)~~ [gscam](https://github.com/ros-drivers/gscam)  |
| [ros-triton-classifier](https://github.com/WHOIGit/ros-triton-classifier) | not ported  |
| [seatrac](https://github.com/WHOIGit/seatrac-api) | not ported  |
| **THIRD-PARTY PACKAGES** | |
| [ds_base](https://bitbucket.org/whoidsl/ds_base/)     | ✅ succeeded by [wr2_base](https://git.whoi.edu/whoi-ros/wr2_base), consider #81  |
| [foxglove_msgs](https://github.com/foxglove/foxglove-sdk) | ✅ |
| [gpsd_client](https://github.com/swri-robotics/gps_umd) | ✅ |
| [rosbridge_suite](https://github.com/RobotWebTools/rosbridge_suite) / [ros-foxglove-bridge](https://github.com/foxglove/ros-foxglove-bridge) | ✅ |

✅ = Fully ported or replaced under ROS 2  
✳️ = Running on ROS 2 with the rospy compatibility layer

rgov committed:
Install ROS1 packages into the ros1 workspace

rgov commented:
In this first round of changes, I am just tweaking the Dockerfile and prepping the ROS1 packages to move into their own `ros1_ws` namespace. I updated the `phyto-arm` script to be aware of this new namespace. Is there anywhere else that needs to be updated?

I have a prototype of `rbr_maestro3_ctd` and of building `ros1_bridge`.

figuernd commented:
If we move to a common node implementation pattern, this may be the time to move towards approaching PhytO-ARM as a service that runs by default, with operations being triggered by service calls rather than launch commands. This is basically how CUREE does things, which makes it possible to operate the vehicle via anything that can make a service call (CLI, Python, or web GUI).

rgov commented:
- [x] Update devcontainer.json to mount the correct paths

rgov committed:
Add install directives to Python nodes

rgov commented:
I now have a [ROS1-ROS2 bridge Dockerfile](https://github.com/rgov/ros1-bridge-container). It will add ~100 lines to our Dockerfile and significantly bloat our container image by a bit since we need to have both ROS versions installed. And, it takes forever to build, so we need to be thoughtful about caching.

figuernd reviewed (approved):
Difficult for me to judge the impact of these changes without running it all on an IFCB and trying out a few configurations - trusting it's well tested I see no issues.

rgov committed:
Force setuptools to use distro's distutils

On the Ubuntu/Debian-packaged Python distribution, distutils is patched to
understand an option `setup.py install --install-layout=deb`. This option is
used by catkin setup.py scripts.

When we installed a newer setuptools (830f301), it came with a newer, vendored
version of distutils which does not have the Ubuntu/Debian patch, and hence
doesn't work with catkin.

By setting the SETUPTOOLS_USE_DISTUTILS=stdlib environment variable, we force
setuptools to use the distro-packaged distutils with the necessary feature.

This was never a problem before when we ran `catkin build` because it did not
actually invoke the `install` command, it did some dry run shenanigans.

Python version constraint: distutils was removed from Python's stdlib in
version 3.12 (PEP 632). This means we cannot upgrade past Python 3.11 while
continuing to support ROS Noetic. (Ubuntu 24.04 ships Python 3.12.)

rgov committed:
Use fix branch for ds_sensor_msgs install bug

https://bitbucket.org/whoidsl/ds_msgs/pull-requests/40

rgov committed:
Migrate multi-stage build with install space

rgov committed:
Major refactor of Dockerfile for simpler dependency management

rgov committed:
Add hot-patch overlay workspace and new container entrypoint

rgov committed:
Add documentation for hot-patching packages in containers

rgov committed:
Modernize CI workflow and analyze with Zizmor

rgov committed:
Initial support for ROS2 and ros1_bridge

rgov commented:
Oof. The CI job took so long to build the ros1_bridge that it timed out after *six hours*.

rgov committed:
Disable external dependencies for ROS2 migration

- Remove ros-rtsp-camera, ros-triton-classifier, seatrac-api from ros1-deps.rosinstall
- Remove ds_util_nodes, seatrac, triton_classifier, rtsp_camera from phyto_arm deps
- Comment out classifier and camera groups in main.launch
- Remove ds_util_nodes/bridge_node from arm_ifcb.launch and arm_seatrac.launch
- Remove ros-launchpad and external packages from Dockerfile build

rgov committed:
Add rospy_too compatibility layer

rgov committed:
Add wr2_msgs as successor to ds_core_msgs and ds_sensor_msgs

rgov commented:
This latest patch series migrates our first-party Python ROS nodes to my [rospy compatibility layer](https://github.com/rgov/rospy-compat). In short, the compatibility layer implements the ROS1 `rospy` and `actionlib` APIs as wrappers around `rclpy`, the Python client library for ROS2. This allows code to run virtually unmodified under ROS2 as a temporary, transitional step.

External packages are not yet addressed, including the critical I/O nodes.

The next step would be incrementally replacing the `rospy` calls with equivalent `rclpy` and dropping the compatibility layer.

rgov committed:
Migrate dli_power_switch package to ROS2

rgov committed:
Migrate jvl_motor package to ROS2

rgov committed:
Migrate ifcb package to ROS2

rgov committed:
Migrate aml_ctd package to ROS2

rgov committed:
Migrate rbr_maestro3_ctd package to ROS2

rgov committed:
Migrate phyto_arm package to ROS2

rgov committed:
Switch to ROS2 Humble and drop ROS1

rgov committed:
Enable wr2_asio adapter_node for serial communications

Now that we're on ROS2 Humble, enable the wr2_asio dependencies
and adapter_node references that replace ds_util_nodes/bridge_node.

rgov committed:
Re-enable rosbridge for ROS 2

rgov committed:
Add a trampoline script that loads our ROS environment

rgov committed:
Use container build stages for ROS package build parallelism

rgov committed:
Replace ros-rtsp-camera with gscam

rgov committed:
Use symlinked install space for easier development

rgov commented:
Note: I abandoned the `rzg/install` branch but since these changes follow it, those commits will still appear in history. I tried to rebase to drop them but it was too difficult.

rgov committed:
Prevent phyto-arm script from dying after docker exec'd child exits

rgov committed:
Re-enable gpsd_client dependency

rgov committed:
Avoid type coercion in AML node

rgov committed:
Fix topic names for ROS 2 (~x -> ~/x)

rgov committed:
Update IFCB node for ROS 2
